src/Storages/Kafka/KeeperHandlingConsumer.h

```diff
@@ -0,0 +1,196 @@
+
+    /// The concept of `prepareToPoll` and `poll` is quite a bit quirky, but I didn't find a better way to:
+    ///   1. Separate the logic of converting messages to rows with virtual columns and everything else
+    ///   2. Do not expose KafkaConsumer2, so KeeperHandlingConsumer has some kind of control over the offsets
+    ///   3. Only do the heavy initialization once the consumer can actually poll messages
+    /// 1: KeeperHandlingConsumer shouldn't be aware of the logic how a single message is turned into rows. It doesn't need to know about virtual columns, error streams, etc.
+    /// 2: By polling `KafkaConsumer2` directly the user (StorageKafka2) should report back the intent size in order to be able to save it before pushing the block. Yes, the
+    ///    the current solution does not solve this (in MessageSinkFunction KeeperHandlingConsumer has no control over the blocks) completely, but I think it is a good compromise.
+    ///    In the future we might add a similar class than `KafkaSource` to handle the conversion of messages to rows, but right now this is not the priority.
+    /// 3: There are several reasons why KeeperHandlingConsumer might not be able to poll messages:
+    ///      - Keeper session has ended
+    ///      - Couldn't get the list of topic partitions from Kafka
+    ///      - Couldn't lock any topic partitions
+    ///    In case of these reasons it doesn't make sense to set up the whole pipeline, which not the lightest operation.
+    /// As a result we have the the MessageSink callback. In the end I think the extra complexity does worth it.
+    std::optional<CannotPollReason> prepareToPoll();
+
+    /// Returns empty optional if polling it not possible or no messages were polled.
+    /// Returns an OffsetGuard that rollbacks the offsets unless it is committed or deactivated.
+    std::optional<OffsetGuard> poll(MessageSinkFunction & message_sink);
+
+    using ConfigCreator = std::function<cppkafka::Configuration(IKafkaExceptionInfoSinkPtr)>;
+    /// Returns true if a new consumer was created
+    bool startUsing(const ConfigCreator & config_creator);
+    void stopUsing();
+    bool isInUse() const { return in_use; }
+    UInt64 getLastUsedUsec() const { return last_used_usec; }
+    bool hasConsumer() const { return kafka_consumer->hasConsumer(); }
+    CppKafkaConsumerPtr moveConsumer() const;
+
+    /// Can be called simultaneously with most of the functions, see the class comment for details.
+    StorageKafkaUtils::ConsumerStatistics getStat() const;
+    IKafkaExceptionInfoSinkPtr getExceptionInfoSink() const;
+private:
+    using TopicPartition = KafkaConsumer2::TopicPartition;
+    using TopicPartitions = KafkaConsumer2::TopicPartitions;
+    using TopicPartitionOffset = KafkaConsumer2::TopicPartitionOffset;
+    using TopicPartitionOffsets = KafkaConsumer2::TopicPartitionOffsets;
+
+    struct LockedTopicPartitionInfo
+    {
+        zkutil::EphemeralNodeHolderPtr lock;
```

Comment:
This seems unused. Not sure if you just meant to have it here because that's the way this lock works. But if it's really protecting something, it might be useful to add TSA annotations to show what's protecting. Feel free to check out `UniqueLock.h` to see how to make a TSA-compatible wrapper.